### PR TITLE
overlord: add barebone fdemgr

### PR DESCRIPTION
**Pull request by valentindavid**

### CODE REVIEW

File: overlord/fdestate/fdemgr.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func (m *FDEManager) Ensure() error {
+	return nil
+}
+
+func getManager(st *state.State) (*FDEManager, error) {
+	c := st.Cached(fdeManagerKey{})
+	if c == nil {
+		return nil, fmt.Errorf("no FDE manager found")
```
Comment (pedronis) [inline overlord/fdestate/fdemgr.go:53]: we usually panic in this cases with internal error, see other helper like this

File: overlord/fdestate/fdemgr.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+package fdestate
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+type FDEManager struct {
```
Comment (pedronis) [inline overlord/fdestate/fdemgr.go:28]: needs a doc comment

File: overlord/fdestate/fdemgr.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package fdestate
```
Comment (pedronis) [inline overlord/fdestate/fdemgr.go:20]: the package itself also needs a doc comment

File: overlord/overlord.go [hunk 4]
```diff
@@ -188,6 +190,8 @@ func New(restartHandler restart.Handler) (*Overlord, error) {
 	// the shared task runner should be added last!
 	o.stateEng.AddManager(o.runner)
 
+	o.addManager(fdestate.Manager(s, o.runner))
```
Comment (pedronis) [inline overlord/overlord.go:193]: probably want to set this up before DeviceManager

File: overlord/fdestate/fdemgr.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+func getManager(st *state.State) (*FDEManager, error) {
+	c := st.Cached(fdeManagerKey{})
+	if c == nil {
+		return nil, fmt.Errorf("no FDE manager found")
+	}
+	manager := c.(*FDEManager)
+	if manager == nil {
+		return nil, fmt.Errorf("FDE manager found has wrong type")
```
Comment (pedronis) [inline overlord/fdestate/fdemgr.go:57]: same about panic

**Review by pedronis — COMMENTED** [review]
did a first pass

File: overlord/fdestate/fdemgr.go [hunk 6]
```diff
@@ -17,6 +17,8 @@
  *
  */
 
+// Package fdestate implements the manager and state responsible for
+// managing full disk encryption keys
```
Comment (pedronis) [inline overlord/fdestate/fdemgr.go:21]: missing . at the end

File: overlord/fdestate/fdemgr.go [hunk 7]
```diff
@@ -25,6 +27,8 @@ import (
 	"github.com/snapcore/snapd/overlord/state"
 )
 
+// FDEManager is responsible for managing managing full disk
+// encryption keys
```
Comment (pedronis) [inline overlord/fdestate/fdemgr.go:29]: same

File: overlord/fdestate/fdemgr.go [hunk 8]
```diff
@@ -51,15 +49,15 @@ func (m *FDEManager) Ensure() error {
...
+func getManager(st *state.State) *FDEManager {
 	c := st.Cached(fdeManagerKey{})
 	if c == nil {
-		return nil, fmt.Errorf("no FDE manager found")
+		panic("internal error: FDE manager is not yet associated with state")
 	}
 	manager := c.(*FDEManager)
 	if manager == nil {
-		return nil, fmt.Errorf("FDE manager found has wrong type")
+		panic("internal error: FDE manager assiciated with the state has the wrong type")
```
Comment (pedronis) [inline overlord/fdestate/fdemgr.go:59]: typo associated

**Review by pedronis — COMMENTED** [review]
thanks, few small comments

**Review by pedronis — APPROVED** [review]
thanks

File: overlord/fdestate/fdemgr.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+	c := st.Cached(fdeManagerKey{})
+	if c == nil {
+		panic("internal error: FDE manager is not yet associated with state")
+	}
+	manager := c.(*FDEManager)
+	if manager == nil {
+		panic("internal error: FDE manager associated with the state has the wrong type")
+	}
+
+	return manager
```
Comment (MiguelPires) [inline overlord/fdestate/fdemgr.go:62]: Go will already panic if the type assertion fails so the code inside the `if` can never run.
```suggestion
	return c.(*FDEManager)
```

File: overlord/fdestate/fdemgr.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+// FDEManager is responsible for managing managing full disk
+// encryption keys.
+type FDEManager struct {
+	state *state.State
+}
+
+type fdeManagerKey struct{}
```
Comment (pedronis) [inline overlord/fdestate/fdemgr.go:34]: we tend to call these types fdeMgrKey

File: overlord/fdestate/fdemgr.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+	st.Cache(fdeManagerKey{}, m)
+
+	return m
+}
+
+func (m *FDEManager) Ensure() error {
+	return nil
+}
+
+func getManager(st *state.State) *FDEManager {
```
Comment (pedronis) [inline overlord/fdestate/fdemgr.go:52]: and this kind of helper: fdeMgr

**Review by pedronis — COMMENTED** [review]
some naming nitpicks

**Review by MiguelPires — APPROVED** [review]
thank you

**Pull request merged**